test/gtest/ucp/test_ucp_am.cc

```diff
@@ -692,6 +692,111 @@ class test_ucp_am_nbx : public test_ucp_am_base {
     ucp_mem_h                       m_rx_memh;
 };
 
+class test_ucp_am_id : public test_ucp_am_nbx {
+public:
+    test_ucp_am_id()
+    {
+        reset_counters();
+    }
+
+protected:
+
+    void reset_counters()
+    {
+        test_ucp_am_nbx::reset_counters();
+        m_recv_counter_cb_1 = 0;
+        m_recv_counter_cb_2 = 0;
+    }
+
+    ucs_status_t set_am_id_overflow_data_handler(entity &e, unsigned am_id,
+                               ucp_am_recv_callback_t cb)
+    {
+        ucp_am_handler_param_t param;
+
+        /* Initialize Active Message data handler */
+        param.field_mask = UCP_AM_HANDLER_PARAM_FIELD_ID |
+                           UCP_AM_HANDLER_PARAM_FIELD_CB |
+                           UCP_AM_HANDLER_PARAM_FIELD_ARG;
+        param.id         = am_id;
+        param.cb         = cb;
+        param.arg        = this;
+
+        return ucp_worker_set_am_recv_handler(e.worker(), &param);
+    }
+
+    void test_am_id_handler(size_t size, size_t header_size = 0ul)
+    {
+        reset_counters();
+
+        EXPECT_EQ(UCS_OK, set_am_id_overflow_data_handler(receiver(),
+                                        0x1, am_id_overflow_data_cb_1));
+        EXPECT_NE(UCS_OK, set_am_id_overflow_data_handler(receiver(),
+                                        0xffff0001, am_id_overflow_data_cb_2));
+
+        ucp_request_param_t param;
+        param.op_attr_mask = 0;
+
+        m_send_counter++;
+        ucs_status_ptr_t sptr = ucp_am_send_nbx(sender().ep(), 0x1, NULL,
+                                                0ul, NULL, 0, &param);
+        wait_receives();
+        EXPECT_EQ(UCS_OK, request_wait(sptr));
+        EXPECT_EQ(m_recv_counter_cb_1, 1);
+        EXPECT_EQ(m_recv_counter_cb_2, 0);
+
+        m_send_counter++;
+        sptr = ucp_am_send_nbx(sender().ep(), 0xffff0001,
+                               NULL, 0ul, NULL, 0, &param);
+        EXPECT_EQ(UCS_PTR_STATUS(sptr), UCS_ERR_INVALID_PARAM);
+        EXPECT_EQ(m_recv_counter_cb_1, 1);
+        EXPECT_EQ(m_recv_counter_cb_2, 0);
+    }
+
+    virtual ucs_status_t am_id_overflow_data_handler(
+                                    const void *header, size_t header_length,
+                                    void *data, size_t length,
+                                    const ucp_am_recv_param_t *rx_param,
+                                    int handler_num)
+    {
+        test_ucp_am_nbx::check_header(header, header_length);
+        m_recv_counter++;
+
+        bool has_reply_ep = get_send_flag();
+
+        EXPECT_EQ(has_reply_ep,
+                  !!(rx_param->recv_attr & UCP_AM_RECV_ATTR_FIELD_REPLY_EP));
+        EXPECT_EQ(has_reply_ep, rx_param->reply_ep != NULL);
+
+        if (1 == handler_num) m_recv_counter_cb_1++;
+        else if (2 == handler_num) m_recv_counter_cb_2++;
+
+        return UCS_OK;
+    }
+
+    static ucs_status_t am_id_overflow_data_cb_1(void *arg, const void *header,
```

Comment:
Code repetition.
Maybe use lambda or `std::bind` to eliminate  it?

Reply:
@gleon99 I was trying to use either lambda or `std::bind` to define a function of type `ucp_am_recv_callback_t` but it didn't work and became too messy.. I am pushing the code without this change for now, but would like to talk about it, maybe I have some gaps in this matter

Reply:
Per our f2f discussion, lambda or `std::bind` won't work here because of C/C++ types interoperability limitation.